erpnext/stock/doctype/serial_no/serial_no.py

```diff
@@ -64,6 +64,37 @@ def __init__(self, *args, **kwargs):
 		super(SerialNo, self).__init__(*args, **kwargs)
 		self.via_stock_ledger = False
 
+	def autoname(self):
+		if frappe.db.get_single_value("Stock Settings", "allow_duplicate_serial_nos"):
+			self.validate_duplicate_in_same_item()
+			self.name = self.autoname_duplicate_serial_no()
+		else:
+			self.name = self.serial_no
+
+	def autoname_duplicate_serial_no(self):
+		"""
+		If duplicate serial no already exists in another item
+		and this behaviour is permitted, append a number to the name to maintain uniqueness.
+		If not, return the name as is.
+		"""
+		duplicate_count_in_other_items = frappe.db.count(
+			"Serial No", {"serial_no": self.serial_no, "item_code": ("!=", self.item_code)}
+		)
+		if duplicate_count_in_other_items > 0:
+			return f"{self.serial_no}-{duplicate_count_in_other_items}"
+		else:
+			return self.name
+
+	def validate_duplicate_in_same_item(self):
+		"""Check to make sure duplicates are not allowed in the same item."""
+		if frappe.db.exists("Serial No", {"serial_no": self.serial_no, "item_code": self.item_code}):
+			frappe.throw(
+				msg=_("Serial No {0} already exists for Item {1}").format(
+					frappe.bold(self.serial_no), frappe.bold(self.item_code)
+				),
+				title=_("Duplicate"),
+			)
```

Comment:
While validation is fine, this should be a DB level unique constraint using `db.add_unique` we have had many such problem with `Bin` where ORM level validations are not sufficient.